### ISSUE TITLE
read's -n option seems to read from line-buffered stdin

### ISSUE THREAD

**Issue by Crestwave** [oilshell/oil#274]
While this would be a nice feature for another option (I had to come up with a [hack](https://github.com/Crestwave/brainbash/blob/master/brainbash#L44) to emulate this on Bash), it's unbuffered on Bash, which allows you to make TUIs and such without having to press enter every time you input something.

Also, have you just not gotten around to implementing the other options, or did you not think they were used? If it's the latter, I've seen `-s`, `-N`, `-p`, and `-d` used numerous times.

**Comment by andychu** [comment]
The buffering might change when we move to native code.

I'm not entirely clear on the problem from your description.  This is an issue for interactive shell programs?  Can you a short code snippet that demonstrates it?

-----

As for the second question, I'm doing things in a "demand-driven" fashion.  Basically a feature is important if we can put an important shell script on this page :)

https://github.com/oilshell/oil/wiki/Shell-Programs-That-Run-Under-OSH

So if you see a shell script that uses those flags, please post it here and that will make it a higher priority.  There's a very long tail of bash features, so the concrete use cases help prioritize work.

I also accept pull requests, because I figure if someone spent the time to implement it, then it must be important to them :)

**Comment by Crestwave** [comment]
> The buffering might change when we move to native code.
> 
> I'm not entirely clear on the problem from your description. This is an issue for interactive shell programs? Can you a short code snippet that demonstrates it?

Something like:

```
while true; do
    read -rn 3
    case $REPLY in
        # Left arrow
        $'\e[C') (( x++ )) ;;
        # Right arrow
        $'\e[D') (( x-- )) ;;
    esac
    printf '\r\e[K%*s' "$x" "A"
```
On Osh, you'd have to press enter for your arrow keys to take effecr. This code snippet would usually use the `-s` flag instead of `stty -echo`, but Osh doesn't have it.

> As for the second question, I'm doing things in a "demand-driven" fashion. Basically a feature is important if we can put an important shell script on this page :)
> 
> https://github.com/oilshell/oil/wiki/Shell-Programs-That-Run-Under-OSH
> 
> So if you see a shell script that uses those flags, please post it here and that will make it a higher priority. There's a very long tail of bash features, so the concrete use cases help prioritize work.

`fff` is actually already listed (the terminal file manager) and uses all of this except `-N`. Any pure Bash TUI would use several of these flags, e.g. https://github.com/dylanaraps/birch, my own https://github.com/Crestwave/shmenu (this uses `-N`; would you want me to post this for that one flag?), https://github.com/dylanaraps/pxltrm, etc.

**Comment by jyn514** [comment]
For comparison, here's an strace of osh and bash respectively:

```
read(0, "\r", 1)                        = 1
write(1, "\n", 1)                       = 1
ioctl(0, TCGETS, {B38400 opost isig -icanon -echo ...}) = 0
ioctl(0, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
read(0, "a", 1)                         = 1
```
```
read(0, "\r", 1)                        = 1
write(2, "\n", 1)                       = 1
ioctl(0, TCGETS, {B38400 opost isig -icanon -echo ...}) = 0
ioctl(0, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
# lots of rtsigaction calls
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig -icanon echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig -icanon echo ...}) = 0
read(0, "a", 1)                         = 1
```

It looks like this is an `ioctl` call, I'm taking a stab at it.

**Comment by andychu** [comment]
Fixed with 0.7.pre2 https://www.oilshell.org/release/0.7.pre2/
